File: py_src/taskito/dashboard/handlers/middleware.py
"""Middleware discovery + per-task enable/disable endpoints."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from taskito.dashboard.errors import _BadRequest, _NotFound

if TYPE_CHECKING:
    from taskito.app import Queue


def handle_list_middleware(queue: Queue, _qs: dict) -> list[dict[str, Any]]:
    """Return every registered middleware with its scopes."""
    return queue.list_middleware()


def handle_get_task_middleware(queue: Queue, _qs: dict, task_name: str) -> dict[str, Any]:
    """Return the middleware chain that fires for ``task_name`` with each
    entry's enabled/disabled state."""
    chain = queue._get_middleware_chain(task_name)
    disabled = set(queue.get_disabled_middleware_for(task_name))
    # Build the full would-fire chain INCLUDING disabled entries so the UI
    # can render every toggle.
    base_chain = queue._global_middleware + queue._task_middleware.get(task_name, [])
    entries: list[dict[str, Any]] = []
    chain_names = {getattr(mw, "name", "") for mw in chain}
    for mw in base_chain:
        name = getattr(mw, "name", "") or f"{type(mw).__module__}.{type(mw).__qualname__}"
        entries.append(
            {
                "name": name,
                "class_path": f"{type(mw).__module__}.{type(mw).__qualname__}",
                "disabled": name in disabled,
                "effective": name in chain_names,
            }
        )
    return {"task": task_name, "middleware": entries}


def handle_put_task_middleware(queue: Queue, body: dict, ids: tuple[str, str]) -> dict[str, Any]:
    task_name, mw_name = ids
    if not isinstance(body, dict) or "enabled" not in body:
        raise _BadRequest('body must include {"enabled": bool}')
    if not isinstance(body["enabled"], bool):
        raise _BadRequest("'enabled' must be a boolean")
    # Confirm the middleware exists in the relevant chain so a typo doesn't
    # silently write a no-op disable entry.
    base_chain = queue._global_middleware + queue._task_middleware.get(task_name, [])
    names = {getattr(mw, "name", "") for mw in base_chain}
    if mw_name not in names:
        raise _NotFound(f"middleware '{mw_name}' is not registered on task '{task_name}'")
    if body["enabled"]:
        new = queue.enable_middleware_for_task(task_name, mw_name)
    else:
        new = queue.disable_middleware_for_task(task_name, mw_name)
    return {"task": task_name, "disabled": new}


def handle_delete_task_middleware(queue: Queue, task_name: str) -> dict[str, bool]:
    """Clear ALL disables for a task — every middleware fires again."""
    return {"cleared": queue.clear_middleware_disables(task_name)}

File: sdks/node/src/dashboard/handlers/middleware.ts
// Middleware discovery + per-task enable/disable endpoints.

import type { Queue } from "../../index";
import { BadRequestError, NotFoundError } from "../errors";

export function listMiddleware(queue: Queue) {
  return queue.listMiddleware();
}

/** The middleware chain for a task with each entry's enabled state. */
export function getTaskMiddleware(queue: Queue, taskName: string) {
  const disabled = new Set(queue.getDisabledMiddlewareFor(taskName));
  const entries = queue.listMiddleware().map((mw) => ({
    name: mw.name,
    class_path: mw.class_path,
    disabled: disabled.has(String(mw.name)),
    effective: !disabled.has(String(mw.name)),
  }));
  return { task: taskName, middleware: entries };
}

export function putTaskMiddleware(
  queue: Queue,
  taskName: string,
  middlewareName: string,
  body: unknown,
) {
  const enabled = (body as { enabled?: unknown } | undefined)?.enabled;
  if (typeof enabled !== "boolean") {
    throw new BadRequestError('body must include {"enabled": bool}');
  }
  // Confirm the middleware exists so a typo can't write a no-op disable entry.
  const names = new Set(queue.listMiddleware().map((mw) => String(mw.name)));
  if (!names.has(middlewareName)) {
    throw new NotFoundError(
      `middleware '${middlewareName}' is not registered on task '${taskName}'`,
    );
  }
  const disabled = enabled
    ? queue.enableMiddlewareForTask(taskName, middlewareName)
    : queue.disableMiddlewareForTask(taskName, middlewareName);
  return { task: taskName, disabled };
}

/** Clear ALL disables for a task — every middleware fires again. */
export function deleteTaskMiddleware(queue: Queue, taskName: string) {
  return { cleared: queue.clearMiddlewareDisables(taskName) };
}
